### ISSUE TITLE
Noux: basic syscall profiling

### ISSUE THREAD

**Issue by cproc** [genodelabs/genode#508]
f3eaacd23784e72d5bb0016f44d74e76b47cafbc uses the x86 time stamp counter to collect some basic syscall execution time statistics. The current results get shown whenever a child exits.

Example report for the non-interactive 'noux_shell_script' test:

```
[init -> noux] quota: avail=244733292, used=2351104
[init -> noux] tsc count total: 19698590817
[init -> noux] tsc count for WAIT4:  12876625062 (~65%)
[init -> noux] tsc count for FORK:   3949616421 (~20%)
[init -> noux] tsc count for EXECVE:     2428466409 (~12%)
[init -> noux] tsc count for READ:   122191398 (~0%)
[init -> noux] tsc count for CLOCK_GETTIME:  114880194 (~0%)
[init -> noux] tsc count for STAT:   76353480 (~0%)
[init -> noux] tsc count for OPEN:   40983894 (~0%)
[init -> noux] tsc count for DIRENT:     37287729 (~0%)
[init -> noux] tsc count for CLOSE:  17143740 (~0%)
[init -> noux] tsc count for FSTAT:  11314377 (~0%)
[init -> noux] tsc count for PIPE:   10061028 (~0%)
[init -> noux] tsc count for WRITE:  8479368 (~0%)
[init -> noux] tsc count for IOCTL:  2339712 (~0%)
[init -> noux] tsc count for FCNTL:  1099638 (~0%)
[init -> noux] tsc count for USERINFO:   648576 (~0%)
[init -> noux] tsc count for LSEEK:  392220 (~0%)
[init -> noux] tsc count for DUP2:   371574 (~0%)
[init -> noux] tsc count for GETTIMEOFDAY:   212247 (~0%)
[init -> noux] tsc count for GETPID:     123750 (~0%)
[init -> noux] tsc count for UTIMES:     0 (~0%)
[init -> noux] tsc count for CONNECT:    0 (~0%)
[init -> noux] tsc count for SHUTDOWN:   0 (~0%)
[init -> noux] tsc count for GETPEERNAME:    0 (~0%)
[init -> noux] tsc count for RECVFROM:   0 (~0%)
[init -> noux] tsc count for RECV:   0 (~0%)
[init -> noux] tsc count for SENDTO:     0 (~0%)
[init -> noux] tsc count for SEND:   0 (~0%)
[init -> noux] tsc count for LISTEN:     0 (~0%)
[init -> noux] tsc count for BIND:   0 (~0%)
[init -> noux] tsc count for ACCEPT:     0 (~0%)
[init -> noux] tsc count for SETSOCKOPT:     0 (~0%)
[init -> noux] tsc count for GETSOCKOPT:     0 (~0%)
[init -> noux] tsc count for SOCKET:     0 (~0%)
[init -> noux] tsc count for SYMLINK:    0 (~0%)
[init -> noux] tsc count for MKDIR:  0 (~0%)
[init -> noux] tsc count for RENAME:     0 (~0%)
[init -> noux] tsc count for READLINK:   0 (~0%)
[init -> noux] tsc count for UNLINK:     0 (~0%)
[init -> noux] tsc count for SELECT:     0 (~0%)
[init -> noux] tsc count for FTRUNCATE:  0 (~0%)
[init -> noux] tsc count for LSTAT:  0 (~0%)
```

**Comment by cproc** [comment]
@alex-ab: I hope it still applies ;-)

**Comment by alex-ab** [comment]
@cproc: Thanks. No chance that this still applies, but it was doable manually ;-).

**Comment by ehmry** [comment]
I made my own profiler, and then @cnuke pointed me to this issue, so we happen to have two independant implementations that look quite similar.

I'm working on a VFS profiler, but I hear that stat caching is known to be not much of a speedup?

**Comment by alex-ab** [comment]
Some time ago I wanted to spend some time in speeding up noux and played with @cproc commit a bit (mainly updating his commit and getting a fs interface which i can call from inside noux to get the statistic data). However, I lost the focus for that and now I'm not using the commits seriously. From my side - feel free to incorporate or to replace the work here with yours. Regarding the stat cache - afair there was no measurable speedup.

**Comment by nfeske** [comment]
The stat cache had actually a positive effect (it reduced the runtime the noux_tool_chain test by a couple of seconds) but it was less than we hoped for. In relation to the performance difference to compiling core natively on Linux (Noux required approximately double the time), the impact of the stat cache was marginal. The run time is apparently overshadowed by another effect. Once that is fixed, we should consider the stat cache again. I have a couple of ideas where the problem may lie:

* I assume that the compiler makes heavy use of libc functions (like string functions). When I ported the libc from FreeBSD, I played it safe and used generic (C-implemented) versions of functions wherever they were available. For many libc functions, there exist architecture-specific assembly versions, that are usually employed on FreeBSD but remain unused on Genode. It may be worthwhile to change our libc in order to use the optimized assembly functions instead of the generic ones.

* Currently, each read operation of a Noux process results in a Noux-session-RPC call. It would probably be good to have a read-ahead buffer locally in the Noux process instead. This buffer could be sensibly dimensioned according to typical block-requests sizes to, let's say, 4 KiB or 8 KiB. So small sporadic reads won't result in much overhead at the file-system and block level. But small continuous reads (like a loop where read is repeatedly called for just a few bytes) would receive a huge speed up. Admittedly, however, I don't know whether the compiler or the build systems shows such read patterns. But the direction could be worth investigating.

* The current implementation of the Noux session as well as the file system interfaces return only a single directory entry at a time. In principle, the reading from the directory handle (file-system session) or the DIRENT sysio operation on Noux could return a while batch of entries. Again, I wanted to be on the save side when implementing those parts to keep the code as simple as possible. But I can imagine that this limitation has a severe impact on the performance of the C preprocessor when it scans the include-search paths for all the used header files.

* For each new process, Noux creates two child components, one for the forked child process and one that is created when the child calls `execve`. This procedure involves overhead in comparison to Linux where fork does not eagerly copy the address space. On the other hand, adding a copy-on-write mechanism to noux would spoil the simplicity of the implementation quite a lot (actually @alex-ab experimented with this idea). I would instead look at the other side effects of the child creation. For example, for each new noux process, we open a couple of ROM sessions (`ld.lib.so`, `libc.lib.so`, `libc_noux.lib.so`, and potentially further libraries), three RM sessions (for the component's address space, for the context area, and for the dynamic linker area), a PD session, and a CPU session. Each of those session requests travels from noux through init (and potentially further parents) to core. It would be interesting to investigate strategies to avoid those session requests altogether. For example, we could cache ROM sessions for often used ROM modules locally within noux and hand out the same ROM session again and again. Or we could think of recycling RM sessions of exited processes for new processes (but this is a bit more complicated as we may need to add an RM operation to reset an RM session).

These are just some ideas that are not supported by any measurement so far. It might be instructive to build a small micro benchmark first to foster a systematic optimization approach.

Such a benchmark could entail only a single step of the noux_tool_chain scenario. E.g., we might trace the `execve` calls of the scenario to find good candidates. I would certainly consider the execution of the actual compile step (I think this is the invocation of `cc1` or `cc1plus`). But the preprocessor or a particular find operation could also be of interest. Then we build a noux scenario that executes only this single step. That means we need to populate the file system of this noux instance with all files produced up to this point of the build. Essentially, we call the step (like `cc1`) as the init process of noux. Once the step completes, noux would exit. So this gives us a nice point where we can measure the run time. Since this scenario no longer relies on fork, we can execute it on base-linux, which clears the way to look at it via Linux profiling tools like oprofile.
